test/others/criu-ns/run.py

```diff
@@ -0,0 +1,164 @@
+        if ret != 0:
+            sys.exit(ret)
+        os._exit(0)
+
+    fd_s.close()
+    os.waitpid(pid, 0)
+
+
+def test_dump_and_restore_without_shell_job(restore_detached=False):
+    check_dumpdir()
+
+    with open("running", "w") as file:
+        pass
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    fd_m.close()
+    if pid == 0:
+        os.setsid()
+        os.dup2(fd_s.fileno(), 0)
+        os.dup2(fd_s.fileno(), 1)
+        os.dup2(fd_s.fileno(), 2)
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+
+    cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v", "-t", str(pid),
+           "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    if ret != 0:
+        sys.exit(ret)
+
+    os.unlink("running")
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    if pid == 0:
+        if restore_detached:
+            cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v",
+                   "--restore-detached", "--criu-binary",
+                   CRIU_BIN]
+        else:
+            cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v",
+                   "--criu-binary", CRIU_BIN]
+
+        ret = subprocess.Popen(cmd, start_new_session=True).wait()
+        if ret != 0:
+            sys.exit(ret)
+        os._exit(0)
+
+    fd_m.close()
+    fd_s.close()
+    os.waitpid(pid, 0)
+
+
+def test_dump_and_restore_pidns():
+    check_dumpdir()
+
+
+    def _dumpee():
+        time.sleep(60)
+
+
+    d_process = multiprocessing.Process(target=_dumpee, daemon=True)
+    d_process.start()
+    dumpee_pid = d_process.pid
+
+    cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v", "--shell-job",
+           "-t", str(dumpee_pid), "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    if ret != 0:
+        sys.exit(ret)
+
+
+    def _restore():
+        fd_m, fd_s = create_pty()
+        fd_m.close()
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v", "--shell-job",
+               "--restore-detached", "--criu-binary", CRIU_BIN]
+        ret = subprocess.Popen(cmd).wait()
+        if ret != 0:
+            sys.exit(ret)
+        fd_s.close()
+
+
+    def _redump(pid):
+        fd_m, fd_s = create_pty()
+        fd_m.close()
+        cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v", "--shell-job",
+           "-t", str(pid), "--criu-binary", CRIU_BIN]
+        ret = subprocess.Popen(cmd).wait()
+        if ret != 0:
+            sys.exit(ret)
+
+
+    r_process = multiprocessing.Process(target=_restore)
+    r_process.start()
+    restored_pid = r_process.pid
+
+    rd_process = multiprocessing.Process(target=_redump, args=(restored_pid,))
+    rd_process.start()
+    rd_process.join()
+
+
+if __name__ == "__main__":
+    test_dump_and_restore_with_shell_job()
+    test_dump_and_restore_without_shell_job()
+    test_dump_and_restore_without_shell_job(restore_detached=True)
+    test_dump_and_restore_pidns()
```

Comment:
Could you add a [documentation comment](https://peps.python.org/pep-0008/#documentation-strings) for each of the above functions to describe what the test does?

Reply:
will do.

Reply:
done.